### PR TITLE
Fix UFT for actual training

### PR DESCRIPTION
**Pull request by superleesa**
- [x] add denoiser loss
- [x] update config
- [x] get data (1500 clear and noisy data respectively)
- [x] add structural loss to adversarial loss
  - [ ] total variational loss
  - [x] SSIM (just use the one from torchmetrics)
- [x] train and see if it's working
- [x] update train loop to fix any bugs

### CODE REVIEW

**Comment by superleesa** [comment]
@weeenieee 
@shoko321 

I think the problem with training with only the adversarial loss is that, it tries to modify the noisy image so that it becomes totally different from the noisy image, rather than "denoising" the noisy image, to reduce the loss.

So what we need is some form of loss that constrains the structural dissimilarily of the noisy image and the restored image. Something like:

total los = Adversarial loss + structural similarity loss

We can use MSE between the noisy and the restored image but it does not really reflect what we want to do because it also considers noises / color differences that we do not need. (so it kind of contradicts with what we want to do = denoising) One way to overcome this might to be something like:

total los = 0.8*Adversarial loss + 0.2*MSE

I will try this but we definitely need another loss for structural similarity. I want to hear your ideas/opinions.

**Comment by superleesa** [comment]
maybe we can use only the structural term from SSIM:

$$
L_s = \frac{\sigma_{xy} + C}{\sigma_x\sigma_y + C}
$$

**Comment by superleesa** [comment]
or [total variational loss](https://lightning.ai/docs/torchmetrics/stable/image/total_variation.html)

**Comment by sho-asn** [comment]
The total variational loss seems to be suitable in this case. It says that it can be used to avoid noisy and overly pixelated results, so I think we can try using the loss?  
https://remi.flamary.com/demos/proxtv.html#:~:text=Definition,in%20the%20image%20processing%20litterature.

**Comment by sho-asn** [comment]
Can we actually try each of them and see the result? If we can improve the performance of the model by changing the loss function, we can finish the fine-tuning. However, if we cannot improve the performance, it will be more problematic.. and trying each of them is time-consuming as well right.

**Comment by wiinnie9** [comment]
so, i think we can now try with either the SSIM loss or the total variation loss, there's a repo on the SSIM loss here but I guess it covers all luminance, contrasts and also structure: https://github.com/NRauschmayr/SSIM_Loss. If our focus would be on the structure whether we can obtain images with similar structure, I found one here on feature matching loss: https://github.com/eli5168/improved_gan_pytorch/blob/f93520646f150a19084efd65215b6c082d3e9d20/improved_GAN.py#L248

**Comment by wiinnie9** [comment]
I think we can try the losses proposed above first for now and if not, we could try changing different other loss functions and can also reconstruct the formula for the loss involving different operators, what do you guys think?

**Comment by superleesa** [comment]
@weeenieee 
- yup SSIM loss might be a good idea, but we want to remove the effect of colors / noises
- feature matching loss seems a good idea (actually the original paper has more than 6000 citations!!!)
  - https://arxiv.org/abs/1606.03498
  - but it might take some time to implement this?

**Comment by superleesa** [comment]
i mean implementing the loss itself seems not too difficult. 
1. we just need to split the discrimnator into two parts, one prior an intermediate activation and one after it. 
2. and then we can use the intermediate activation and simply calculate the difference between the activation of the restored and clear images.

but as the authors argue in the paper, naive implementation of this might suffer from mode collapsing so i'm not sure.

**Comment by superleesa** [comment]
actually
- SSIM is robust to noises -> (because computation on local boxes & normalization by indivisual image variance terms) -> noises get ignored (even if noise is present in the restored image, score does not change much)
- SSIM does not really weight color information -> color differences between clear vs. restored will be (kind of) ignored

so SSIM is a fairly good meric.

**Comment by superleesa** [comment]
fix ssim loss -> should have same normalization factors across all sand dust images and clear images
